### PR TITLE
[GPU] Add test utility functions

### PR DESCRIPTION
**Pull request by tuxedcat**
### Details:
 - Add useful utility functions when writing/debugging test.
 - Fix  conv_onednn_cases/convolution_gpu_onednn.conv_onednn_cases/in8x8y1z32f_output32f_filter3x3y1z_groups1_stride1_batch32_format0_with_bias_1_onednn_impl
 - Minor refactoring

### Tickets:
 - 67491

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/network.cpp [hunk 1]
```diff
@@ -759,6 +751,14 @@ layout network::get_node_output_layout(const primitive_id& output_id) const {
...
 }
 
+memory::ptr network::get_output_memory(const primitive_id& id) {
+    return get_primitive(id)->output_memory_ptr();
+}
+
+layout network::get_output_layout(const primitive_id& id) const {
+    return get_primitive(id)->get_output_layout();
+}
+
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/network.cpp:761]: why did you change this?

File: src/plugins/intel_gpu/tests/test_cases/convolution_gpu_test.cpp [hunk 2]
```diff
@@ -9083,11 +9084,11 @@ TEST_P(convolution_gpu_onednn, conv_onednn_cases) {
     for (auto& p : network.get_primitives_info())
         std::cerr << p.original_id << " " << p.kernel_id << std::endl;
 
-    auto out_mem = network.get_output("conv_fsv").get_memory();
-    mem_lock<FLOAT16> out_ptr{ out_mem, get_test_stream() };
-    auto out_lay = out_mem->get_layout();
+    auto out_ptr = network.get_output_values<FLOAT16>("conv_fsv");
+    auto out_lay = network.get_output_layout("conv_fsv");
+    ASSERT_TRUE(is_eq_data_type_T<FLOAT16>(out_lay.data_type));
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/test_cases/convolution_gpu_test.cpp:9089]: why do we need this function? What about just checking `ASSERT_EQ(out_lay.data_type, data_types::f16)`?

File: src/plugins/intel_gpu/tests/test_cases/convolution_gpu_test.cpp [hunk 3]
```diff
@@ -9083,11 +9084,11 @@ TEST_P(convolution_gpu_onednn, conv_onednn_cases) {
...
 
-    ASSERT_EQ(out_mem->get_layout().format, format::byxf);
+    ASSERT_EQ(out_lay.format.to_string(), format(format::byxf).to_string());
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/test_cases/convolution_gpu_test.cpp:9091]: why is it necessary to check as string?

File: src/plugins/intel_gpu/tests/test_cases/convolution_gpu_test.cpp [hunk 4]
```diff
@@ -9071,6 +9071,7 @@ TEST_P(convolution_gpu_onednn, conv_onednn_cases) {
         conv_fsv.output_paddings = {padding({ 0, 0, 0, 0 }, 0.f)};
         topology.add(conv_fsv);
     }
+    topology.add(reorder("reorder_bfyx", input_info("conv_fsv"), format::bfyx, data_types::f32));
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/test_cases/convolution_gpu_test.cpp:9074]: why is `reorder_bfyx` needed?

File: src/plugins/intel_gpu/tests/test_utils/test_utils.h [hunk 5]
```diff
@@ -581,6 +581,33 @@ T div_up(const T a, const U b) {
...
+    for (T i : net.get_output_values<T>(prim_id, max_cnt))
+        std::cout << std::setw(6) << std::to_string(i).substr(0, 5) << ' ';
+}
+template <>
+inline void print_primitive<FLOAT16>(network& net, const std::string& prim_id, size_t max_cnt) {
+    std::cout << std::endl << "==========" << prim_id << "==========" << std::endl;
+    for (float i : net.get_output_values<FLOAT16>(prim_id, max_cnt))
+        std::cout << std::setw(6) << std::to_string(i).substr(0, 5) << ' ';
+}
+
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/test_utils/test_utils.h:610]: This might be useful, but it is not used in the source code itself. Then I think we cannot merge this function.

**Review by isanghao — COMMENTED** [review]
The purpose of utils is a bit blurry. This might be useful for you, but I'm not sure whether it is generally helpful for others too... :(

**Review by isanghao — CHANGES_REQUESTED** [review]
Please see my previous comment.

File: src/plugins/intel_gpu/src/graph/network.cpp (hunk 1 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/src/graph/network.cpp:761]: `output_id` is not proper parameter name for this function

File: src/plugins/intel_gpu/tests/test_cases/convolution_gpu_test.cpp (hunk 3 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/tests/test_cases/convolution_gpu_test.cpp:9091]: It's for more readable error message. Enum comparison prints unhelpful integer value.

File: src/plugins/intel_gpu/tests/test_cases/convolution_gpu_test.cpp (hunk 4 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/tests/test_cases/convolution_gpu_test.cpp:9074]: This test case is in WIP. As far as I know, output node has some special treatment. Therefore I separated convolution node and output node to avoid such special treatment.

File: src/plugins/intel_gpu/tests/test_cases/convolution_gpu_test.cpp (hunk 2 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/tests/test_cases/convolution_gpu_test.cpp:9089]: Reverted related changes.

File: src/plugins/intel_gpu/tests/test_utils/test_utils.h (hunk 5 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/tests/test_utils/test_utils.h:610]: Removed.

File: src/plugins/intel_gpu/tests/test_cases/convolution_gpu_test.cpp (hunk 4 above)
Comment (isanghao) [reply src/plugins/intel_gpu/tests/test_cases/convolution_gpu_test.cpp:9074]: Why did you update the case if it is WIP?

File: src/plugins/intel_gpu/tests/test_cases/convolution_gpu_test.cpp [hunk 6]
```diff
@@ -9083,11 +9084,10 @@ TEST_P(convolution_gpu_onednn, conv_onednn_cases) {
     for (auto& p : network.get_primitives_info())
         std::cerr << p.original_id << " " << p.kernel_id << std::endl;
 
-    auto out_mem = network.get_output("conv_fsv").get_memory();
-    mem_lock<FLOAT16> out_ptr{ out_mem, get_test_stream() };
-    auto out_lay = out_mem->get_layout();
-
-    ASSERT_EQ(out_mem->get_layout().format, format::byxf);
+    auto out_ptr = network.get_output_values<FLOAT16>("conv_fsv");
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/test_cases/convolution_gpu_test.cpp:9087]: Could you explain why it is changed to `get_output_values`?

Comment (tuxedcat) [reply src/plugins/intel_gpu/tests/test_cases/convolution_gpu_test.cpp:9087]: https://github.com/openvinotoolkit/openvino/pull/14738#discussion_r1058041538 As I mentioned before, it is to avoid side-effects related to output nodes. For example, [this](https://github.com/openvinotoolkit/openvino/pull/14837/files#diff-41a32bc38b6fa9c7db5e656050c380c993d424df879bb6a1e2627e7008a7704bR1866) patch is related to the side-effect of output node. Moreover, many other testcases are also using reorder_bfyx as an output node (possibly for the same purpose as above), and there is no reason that convolution node has to be output node in this testcase. Besides, #14860(new PR of #14143) is not WIP now (in review). Therefore I strongly recommend this change for the view of consistency and safety.
https://github.com/openvinotoolkit/openvino/pull/14738#discussion_r1058041916 Because convolution node is not output node anymore. Using get_output_values() is safer (and more readable!) than using memlock because it guarantees correctness of memory data type.

File: src/plugins/intel_gpu/tests/test_cases/convolution_gpu_test.cpp (hunk 4 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/tests/test_cases/convolution_gpu_test.cpp:9074]: For example, [this](https://github.com/openvinotoolkit/openvino/pull/14837/files#diff-41a32bc38b6fa9c7db5e656050c380c993d424df879bb6a1e2627e7008a7704bR1866) patch is related to the side-effect of output node. Moreover, many other testcases are also using reorder_bfyx as an output node (possibly for the same purpose as above), and there is no reason that convolution node has to be output node in this testcase. Besides, #14860(new PR of #14143) is not WIP now (in review). Therefore I strongly recommend this change for the view of consistency and safety.

File: src/plugins/intel_gpu/tests/test_cases/convolution_gpu_test.cpp (hunk 6 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/tests/test_cases/convolution_gpu_test.cpp:9087]: Furthermore, original code throws exception: Unable to lock allocation_type::usm_device with write lock_type.

**Comment by tuxedcat** [comment]
Offline discussed) Removed byxf format forcing because it cause ocl:ref selection instead jit:ir.

File: src/plugins/intel_gpu/include/intel_gpu/graph/network.hpp [hunk 7]
```diff
@@ -125,10 +125,27 @@ struct network {
...
-    memory::ptr get_output_memory(const primitive_id& output_id);
     layout get_node_output_layout(const primitive_id& output_id) const;
-    layout get_output_layout(const primitive_id& output_id) const;
+
+    template <class T, class U = T>
+    std::vector<U> get_output_values(const primitive_id& id, size_t max_cnt = std::numeric_limits<size_t>::max()) {
+        std::vector<U> ret;
+        auto ptr = get_output_memory(id);
+        auto out_ids = get_output_ids();
+        if (find(out_ids.begin(), out_ids.end(), id) == out_ids.end() && get_engine().configuration().use_memory_pool)
```
Comment (isanghao) [inline src/plugins/intel_gpu/include/intel_gpu/graph/network.hpp:135]: Let's remove this condition `get_engine().configuration().use_memory_pool`

File: src/plugins/intel_gpu/include/intel_gpu/runtime/memory.hpp [hunk 8]
```diff
@@ -43,6 +43,7 @@ struct memory {
 
     size_t size() const { return _bytes_count; }
     size_t count() const { return _layout.count(); }
+    bool is_reused() const { return _reused; }
```
Comment (isanghao) [inline src/plugins/intel_gpu/include/intel_gpu/runtime/memory.hpp:46]: we don't need this anymore

**Review by isanghao — CHANGES_REQUESTED** [review]

File: src/plugins/intel_gpu/include/intel_gpu/graph/network.hpp [hunk 9]
```diff
@@ -125,10 +125,27 @@ struct network {
             evt = get_primitive_event(output_id);
         return network_output(evt, get_output_memory(output_id), get_stream_ptr());
     }
-
-    memory::ptr get_output_memory(const primitive_id& output_id);
     layout get_node_output_layout(const primitive_id& output_id) const;
-    layout get_output_layout(const primitive_id& output_id) const;
+
+    template <class T, class U = T>
```
[the remark this answers is not in the document]
Comment (tuxedcat) [reply src/plugins/intel_gpu/include/intel_gpu/graph/network.hpp]: Done.

File: src/plugins/intel_gpu/include/intel_gpu/graph/network.hpp [hunk 10]
```diff
@@ -125,10 +125,27 @@ struct network {
...
+    std::vector<U> get_output_values(const primitive_id& id, size_t max_cnt = std::numeric_limits<size_t>::max()) {
+        std::vector<U> ret;
+        auto ptr = get_output_memory(id);
+        auto out_ids = get_output_ids();
+        if (find(out_ids.begin(), out_ids.end(), id) == out_ids.end() && get_engine().configuration().use_memory_pool)
+            IE_THROW()
+                << "Please set engine_configuration.use_memory_pool option to false to get values of non-output node."
+                << std::endl;
```
[the remark this answers is not in the document]
Comment (tuxedcat) [reply src/plugins/intel_gpu/include/intel_gpu/graph/network.hpp]: Done.

File: src/plugins/intel_gpu/include/intel_gpu/graph/network.hpp (hunk 7 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/include/intel_gpu/graph/network.hpp]: Done.

File: src/plugins/intel_gpu/include/intel_gpu/runtime/memory.hpp (hunk 8 above)
Comment (tuxedcat) [reply src/plugins/intel_gpu/include/intel_gpu/runtime/memory.hpp]: Done.

**Review by isanghao — APPROVED** [review]

File: src/plugins/intel_gpu/include/intel_gpu/graph/network.hpp [hunk 11]
```diff
@@ -128,9 +128,25 @@ struct network {
...
+            IE_THROW() << "Non output node's memory may have been reused. "
+                          "Make target node to output by using ov::intel_gpu::custom_outputs in ExecutionConfig.";
+        mem_lock<T, mem_lock_type::read> mem(ptr, get_stream());
+        if (ptr->get_layout().data_type != type_to_data_type<T>::value)
+            IE_THROW() << "target type " << data_type_traits::name(type_to_data_type<T>::value)
+                       << " mismatched with actual type " << data_type_traits::name(ptr->get_layout().data_type);
+        for (size_t i = 0; i < std::min(max_cnt, ptr->get_layout().count()); i++)
+            ret.push_back(mem[i]);
+        return ret;
+    }
```
Comment (isanghao) [inline src/plugins/intel_gpu/include/intel_gpu/graph/network.hpp:148]: Oops sorry for the belated response, but this function is for test purpose. In that case, it would be better to put it in test code instead of execution code.

**Review by isanghao — CHANGES_REQUESTED** [review]

**Review by isanghao — APPROVED** [review]
